Improve module augmentation in `table` plugin

filipsobol opened the pull request:
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Internal: Improve TypeScript configuration in the project and module augmentation in `table` plugin. Related to #13565.

niegowski commented on packages/ckeditor5-table/tsconfig.json:
```diff
@@ -1,7 +1,8 @@
 {
-  "extends": "ckeditor5/tsconfig.json",
+  "extends": "../../tsconfig.json",
```
Why such change?

filipsobol replied:
To make it consistent with `tsconfig.release.json` in the same package.

niegowski commented on packages/ckeditor5-table/plugin.d.ts:
```diff
@@ -0,0 +1,136 @@
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+import type {
```
Why this is `d.ts`? Maybe this could be a standard `ts` file?

niegowski commented:
I'm also concerned about automatic tests for other packages. Those import other features to test integrations and they probably do not import using `index` files. Those could fail if augmentations are not provided.

filipsobol commented on packages/ckeditor5-table/plugin.d.ts:
```diff
@@ -0,0 +1,136 @@
+    insertTableRowAbove: InsertRowCommand;
+		insertTableRowBelow: InsertRowCommand;
+    insertTable: InsertTableCommand;
+    mergeTableCellRight: MergeCellCommand;
```
Fixed

filipsobol commented on packages/ckeditor5-table/plugin.d.ts:
```diff
@@ -0,0 +1,136 @@
+    [ TableUI.pluginName ]: TableUI;
+    [ TableUtils.pluginName ]: TableUtils;
+		[ PlainTableOutput.pluginName ]: PlainTableOutput;
+	}
```
Fixed

Reinmar commented on packages/ckeditor5-table/src/index.ts:
```diff
@@ -29,3 +29,36 @@ export { default as TableSelection } from './tableselection';
 export { default as TableColumnResizeEditing } from './tablecolumnresize/tablecolumnresizeediting';
+
+export type { TableConfig } from './tableconfig';
+export type { default as InsertColumnCommand } from './commands/insertcolumncommand';
```
Why only export types? AFAIK we usually export entire commands. They are a public API of a feature (both, as registered in commands registry and classes that can be reused).

Reinmar replied:
F2F discussion: It turned out that we didn't expose any commands so far in index.js files. I thought we did expose some already. Hence, let's continue not exposing anything for now (except types if needed).

Reinmar commented on packages/ckeditor5-table/src/plugin.ts:
```diff
@@ -0,0 +1,139 @@
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+import type {
```
This looks like a duplication of index.ts. Do we really need that?

filipsobol replied:
On today's daily, the team agreed to keep module augmentation in a separate `augmentation.ts` file.

Reinmar commented on packages/ckeditor5-table/src/plugin.ts:
```diff
@@ -0,0 +1,139 @@
+/**
```
As I commented in the PR thread, I'd consider renaming this to `package.ts` as it contains kind of metadata for a package. But then I'd merge the `fooconfig.ts` file into this one too. Why having configs in one place and plugin and command maps in another?

filipsobol replied:
On today's daily, the team agreed to use the `augmentation.ts` file name and to store config in a separate file.

Reinmar commented:
My biggest doubt regarding this is: why do we have `plugin.d.ts` and `index.ts`? Why not merging them into one file? And keeping there all the module augmentation too?

Reinmar commented:
Note: Squash all commits in this branch before the merge so to not pollute the history with all the test import changes.

filipsobol commented:
Adding `"./packages/*/src/augmentation.ts"` to `include` array in `tsconfig.json` didn't work when running `testautomated.js`. However, it worked when we added `"./packages/ckeditor5-table/src/augmentation.ts"` to `files`.

When debugging this with @niegowski and comparing behavior of `tsc` and `testautomated.js`, we found that `tsc` worked with both configurations. After we enabled the `logLevel: 'info'` option in `ts-loader` we noticed that it loads `tsconfig.js` for `paragraph` plugin, despite testing `clipboard` plugin in root of the monorepo project. When we added `"./packages/*/src/augmentation.ts"` to `include` array in `./packages/ckeditor5-paragraph/tsconfig.json`, it started working as expected for `clipboard`.

We need to investigate why seemingly random `tsconfig.json` is being loaded by `ts-loader`. For now, as a quick workaround, we will add `"./packages/*/src/augmentation.ts"`to `include` arrays in all packages.

filipsobol commented:
Above issue with the wrong \`tsconfig.json\` being loaded will be fixed in ckeditor/ckeditor5-dev#847

Reinmar commented:
Note to self: I wanted to comment that you changed formatting of JSON files and I remember that we have a special rule for them in `.editorconfig` to use two spaces. But then I checked it's for `package.json`s only :D And we actually use tabs in other JSON files. Case closed :D

niegowski commented:
Please update the "Suggested merge commit message" to address changes in this PR.

pomek commented on tsconfig.test.json:
```diff
@@ -0,0 +1,8 @@
+	"include": [
+		"./typings/",
+		"./packages/*/src/augmentation.ts",
+		"./external/*/packages/*/src/augmentation.ts"
```
```suggestion
		"./external/ckeditor5-internal/packages/*/src/**/*.ts",
		"./external/collaboration-features/packages/*/src/**/*.ts",
```

To match https://github.com/ckeditor/ckeditor5/blob/0e5ddd74e5684897d2d2c87a0e65150c287433d4/tsconfig.docs.json#L5-L6. People can have more than two repositories in `external/` and it should. not be loaded.

Btw, we use spaces in all JSON files.

filipsobol replied:
I updated it to:  
 

```
"./external/ckeditor5-internal/packages/*/src/augmentation.ts",
"./external/collaboration-features/packages/*/src/augmentation.ts"
```

pomek commented on scripts/check-manual-tests.sh:
```diff
@@ -26,6 +27,11 @@ then
   MANUAL_TEST_SERVER_OPTIONS="$MANUAL_TEST_SERVER_OPTIONS -i $IDENTITY_FILE"
 fi
 
+if [ ! -z "$TS_CONFIG" ]
+then
+  MANUAL_TEST_SERVER_OPTIONS="$MANUAL_TEST_SERVER_OPTIONS --tsconfig $TS_CONFIG"
+fi
```
I would add `--tsconfig ./tsconfig.test.json` to L43 and remote support for the `-t` flag. This script is used only for verification on CI and when releasing stuff. We do not have customize it.

pomek reviewed (approved):
LGTM. But TBH, too many changes. I will depend on CI results.
